workers_node_groups.tf

```diff
@@ -0,0 +1,176 @@
+resource "aws_iam_role" "managed_node_groups" {
+  count                 = local.worker_group_managed_node_group_count > 0 ? 1 : 0
+  name                  = "${var.workers_role_name != "" ? var.workers_role_name : aws_eks_cluster.this.name}-managed-node-groups"
+  assume_role_policy    = data.aws_iam_policy_document.workers_assume_role_policy.json
+  permissions_boundary  = var.permissions_boundary
+  path                  = var.iam_path
+  force_detach_policies = true
+  tags                  = var.tags
+}
+
+resource "aws_iam_role_policy_attachment" "managed_node_groups_AmazonEKSWorkerNodePolicy" {
+  count      = local.worker_group_managed_node_group_count > 0 ? 1 : 0
+  policy_arn = "arn:aws:iam::aws:policy/AmazonEKSWorkerNodePolicy"
+  role       = aws_iam_role.managed_node_groups[0].name
+}
+
+resource "aws_iam_role_policy_attachment" "managed_node_groups_AmazonEKS_CNI_Policy" {
+  count      = local.worker_group_managed_node_group_count > 0 ? 1 : 0
+  policy_arn = "arn:aws:iam::aws:policy/AmazonEKS_CNI_Policy"
+  role       = aws_iam_role.managed_node_groups[0].name
+}
+
+resource "aws_iam_role_policy_attachment" "managed_node_groups_AmazonEC2ContainerRegistryReadOnly" {
+  count      = local.worker_group_managed_node_group_count > 0 ? 1 : 0
+  policy_arn = "arn:aws:iam::aws:policy/AmazonEC2ContainerRegistryReadOnly"
+  role       = aws_iam_role.managed_node_groups[0].name
+}
+
+resource "aws_iam_role_policy_attachment" "managed_node_groups_additional_policies" {
+  count      = local.worker_group_managed_node_group_count > 0 ? length(var.workers_additional_policies) : 0
+  role       = aws_iam_role.managed_node_groups[0].name
+  policy_arn = var.workers_additional_policies[count.index]
+}
+
+resource "aws_iam_role_policy_attachment" "managed_node_groups_autoscaling" {
+  count      = var.manage_worker_autoscaling_policy && var.attach_worker_autoscaling_policy && local.worker_group_managed_node_group_count > 0 ? 1 : 0
+  policy_arn = aws_iam_policy.managed_node_groups_autoscaling[0].arn
+  role       = aws_iam_role.managed_node_groups[0].name
+}
+
+resource "aws_iam_policy" "managed_node_groups_autoscaling" {
+  count       = var.manage_worker_autoscaling_policy && local.worker_group_managed_node_group_count > 0 ? 1 : 0
+  name_prefix = "eks-worker-autoscaling-${aws_eks_cluster.this.name}"
+  description = "EKS worker node autoscaling policy for cluster ${aws_eks_cluster.this.name}"
+  policy      = data.aws_iam_policy_document.worker_autoscaling.json
+  path        = var.iam_path
+}
+
+resource "random_pet" "managed_node_groups" {
+  count = local.worker_group_managed_node_group_count
```

Comment:
Every occurrence of `count = local.worker_group_managed_node_group_count` is strong candidate for switching to `for_each = toset(var.workers_additional_policies)`.

Reply:
Can I replace that look up though? I would still need to provide a default value

Reply:
Good point. Given that [node_group_name](https://www.terraform.io/docs/providers/aws/r/eks_node_group.html) is a required argument for the eks_node_group resource it seems reasonable to ask that module users specify it along with things like the max size.

Reply:
```
The given "for_each" argument value is unsuitable: "for_each" supports maps
and sets of strings, but you have provided a set containing type object.
```
Can't use for each on the random_pets

Reply:
Node groups must have unique names, therefore you can use those as an index. Try creating a local variable that maps the name to the dictionary and instead reference that everywhere:

```hcl
locals {
  node_groups = { for obj in toset(var.worker_group_managed_node_groups) : obj.name => obj }
}
```

Reply:
@eytanhanig I can't seem to fix this error when using for_each:
```
Error: Invalid function argument

  on ../../node_groups.tf line 88, in resource "aws_eks_node_group" "workers":
  87: 
  88:       [
  89: 
  90: 
  91: 
  92: 
  93: 
    |----------------
    | count.index is 0
    | random_pet.node_groups is object with 1 attribute "example"
    | var.node_groups is tuple with 1 element

Invalid value for "list" parameter: element 2: string required.
```
Can you take a look at it and see what you can find?

Reply:
Because you're switching from count to for_each, `var.node_groups[count.index]` becomes `each.value`.

If we require that each node_group have a unique name then you can ditch the use of `random_pet`, resulting in this block simplifying even further:

```hcl
resource "aws_eks_node_group" "workers" {
  for_each = local.node_groups

  cluster_name    = var.cluster_name
  node_group_name = join("-", [var.cluster_name, each.key])
  node_role_arn   = lookup(each.value, "iam_role_arn", aws_iam_role.node_groups[0].arn)
  subnet_ids      = lookup(each.value, "subnets", local.workers_group_defaults["subnets"])

  scaling_config {
    desired_size = lookup(each.value, "node_group_desired_capacity", local.workers_group_defaults["asg_desired_capacity"])
    max_size     = lookup(each.value, "node_group_max_capacity", local.workers_group_defaults["asg_max_size"])
    min_size     = lookup(each.value, "node_group_min_capacity", local.workers_group_defaults["asg_min_size"])
  }

  ami_type        = lookup(each.value, "ami_type", null)
  disk_size       = lookup(each.value, "root_volume_size", null)
  instance_types  = [lookup(each.value, "instance_type", null)]
  labels          = lookup(each.value, "node_group_k8s_labels", null)
  release_version = lookup(each.value, "ami_release_version", null)

  remote_access {
    ec2_ssh_key               = lookup(each.value, "key_name", "") != "" ? each.value["key_name"] : null
    source_security_group_ids = lookup(each.value, "key_name", "") != "" ? lookup(
      each.value,
      "source_security_group_ids",
      []
    ) : null
  }

  tags = lookup(each.value, "node_group_additional_tags", null)

  lifecycle {
    create_before_destroy = true
  }
}
```